pennylane/interfaces/execution.py

```diff
@@ -360,6 +360,22 @@ def cost_fn(params, x):
         for i, tape in enumerate(tapes):
             tapes[i] = expand_fn(tape)
 
+        if gradient_kwargs.get("method", "") == "adjoint_jacobian":
+            # adjoint jacobian needs further expansion
+            # this is a quick patch that should be cleaned up by later refactoring
+            if INTERFACE_MAP[interface] == "jax":
+                stop_at = (
+                    ~qml.operation.is_measurement & qml.operation.has_nopar | qml.operation.has_gen
+                )
+            else:
+                stop_at = (
+                    ~qml.operation.is_measurement & ~qml.operation.is_trainable
+                    | qml.operation.has_gen

```

Comment:
Are these meant to be bitwise operations or logical?
If logical, I'd suggest using `not` `and` and `or` instead.

Also, the above can be rewritten as
```python
not (qml.operation.is_measurement or qml.operation.is_trainable)
                    or qml.operation.has_gen
```
to reduce the number of not operations by 1.

Reply:
The point of adding `~qml.operation.is_measurement` is to exit with False early if the object is a measurement.  `qml.operation.is_trainable` will raise an `AttributeError` if called on a measurement.

Simplifying it in the way suggested would give `AttributeError`'s.

Also, see `qml.BooleanFn`.  The class allows `~fn` and `fn1 & fn2`, but not `not fn` or `fn1 and fn2`.

Reply:
As discussed, I think this involves changes to implementation elsewhere as the use of bitwise operations for logical meaning can be dangerous. Will resolve this for now.

Reply:
> Also, see `qml.BooleanFn`. The class allows `~fn` and `fn1 & fn2`, but not `not fn` or `fn1 and fn2`.

iirc we originally wanted to use the logical operators, but I think we ran into trouble in that Python doesn't make it easy to define custom behaviour for them (unlike the bitwise operators).